src/platform/windows/nvprefs/nvprefs_interface.cpp

```diff
@@ -0,0 +1,225 @@
+#include "nvprefs_common.h"
+
+#include "nvprefs_interface.h"
+
+#include "driver_settings.h"
+#include "undo_data.h"
+#include "undo_file.h"
+
+namespace {
+
+  const auto sunshine_program_data_folder = "Sunshine";
+  const auto nvprefs_undo_file_name = "nvprefs_undo.json";
+
+}  // namespace
+
+namespace nvprefs {
+
+  struct nvprefs_interface::impl {
+    bool loaded = false;
+    driver_settings_t driver_settings;
+    std::filesystem::path undo_folder_path;
+    std::filesystem::path undo_file_path;
+    std::optional<undo_data_t> undo_data;
+    std::optional<undo_file_t> undo_file;
+  };
+
+  nvprefs_interface::nvprefs_interface():
+      pimpl(new impl()) {
+  }
+
+  nvprefs_interface::~nvprefs_interface() {
+    if (owning_undo_file() && load()) {
+      restore_global_profile();
+    }
+    unload();
+  }
+
+  bool
+  nvprefs_interface::load() {
+    if (!pimpl->loaded) {
+      // Check %ProgramData% variable, need it for storing undo file
+      wchar_t program_data_env[MAX_PATH];
+      auto get_env_result = GetEnvironmentVariableW(L"ProgramData", program_data_env, MAX_PATH);
```

Comment:
Maybe we should use `%LOCALAPPDATA%` instead of `%ProgramData%` to keep the undo state for installer and portable versions of Sunshine apart.

Reply:
The NVIDIA settings themselves are global and stored in %ProgramData% file. If we split out undo states, one can pollute the global state with no way for the other one to know about it.